joeynmt/transformer.py

```diff
@@ -0,0 +1,233 @@
+# -*- coding: utf-8 -*-
+
+import math
+import copy
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+# This contains helper classes for the Transformer.
+# Source: http://nlp.seas.harvard.edu/2018/04/03/attention.html
+
+
+def clones(module, N):
+    """Produce N identical layers."""
+    return nn.ModuleList([copy.deepcopy(module) for _ in range(N)])
+
+
+# pylint: disable=arguments-differ,too-many-arguments
+class SublayerConnection(nn.Module):
+    """
+    Transformer residual connection
+    Applies sublayer to x, layer norm, dropout, then adds x.
+    """
+    def __init__(self, size, dropout):
+        super(SublayerConnection, self).__init__()
+        self.norm = nn.LayerNorm(size)
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self, x, sublayer):
+        return x + self.dropout(sublayer(self.norm(x)))
+
+
+def subsequent_mask(size):
+    """
+    Mask out subsequent positions.
+    """
+    attn_shape = (1, size, size)
+    subseq_mask = np.triu(np.ones(attn_shape), k=1).astype('uint8')
+    return torch.from_numpy(subseq_mask) == 0
+
+
+def attention(query, key, value, mask=None, dropout=None):
```

Comment:
Can we merge this with our Attention module?

Reply:
Yes we could, but let's leave it here for now.